packages/flutter/lib/src/widgets/two_dimensional_viewport.dart

```diff
@@ -0,0 +1,1216 @@
+    const  String hint = 'Subclasses should call applyContentDimensions on the '
+      'verticalOffset and horizontalOffset to set the min and max scroll offset.';
+    assert(() {
+      if (!(verticalOffset as ScrollPosition).hasContentDimensions) {
+        throw FlutterError.fromParts(<DiagnosticsNode>[
+          ErrorSummary(
+            'The verticalOffset was not given content dimensions during '
+            'layoutChildSequence.'
+          ),
+          ErrorHint(hint),
+        ]);
+      }
+      return true;
+    }());
+    assert(() {
+      if (!(horizontalOffset as ScrollPosition).hasContentDimensions) {
+        throw FlutterError.fromParts(<DiagnosticsNode>[
+          ErrorSummary(
+            'The horizontalOffset was not given content dimensions during '
+            'layoutChildSequence.'
+          ),
+          ErrorHint(hint),
+        ]);
+      }
+      return true;
+    }());
+  }
+
+  /// Returns the child for a given [ChildVicinity].
+  ///
+  /// This method will build the child if it has not been already, or will reuse
+  /// it if it already exists.
+  RenderBox? buildOrObtainChildFor(ChildVicinity vicinity) {
+    assert(vicinity != ChildVicinity.invalid);
+    if (_leadingXIndex == null || _trailingXIndex == null || _leadingXIndex == null || _trailingYIndex == null) {
+      // First child of this layout pass. Set leading and trailing trackers.
+      _leadingXIndex = vicinity.xIndex;
+      _trailingXIndex = vicinity.xIndex;
+      _leadingYIndex = vicinity.yIndex;
+      _trailingYIndex = vicinity.yIndex;
+    } else {
+      // If any of these are still null, we missed a child.
+      assert(_leadingXIndex != null);
+      assert(_trailingXIndex != null);
+      assert(_leadingYIndex != null);
+      assert(_trailingYIndex != null);
+
+      // Update as we go.
+      _leadingXIndex = math.min(vicinity.xIndex, _leadingXIndex!);
+      _trailingXIndex = math.max(vicinity.xIndex, _trailingXIndex!);
+      _leadingYIndex = math.min(vicinity.yIndex, _leadingYIndex!);
+      _trailingYIndex = math.max(vicinity.yIndex, _trailingYIndex!);
+    }
+    if (_needsChildRebuild || !_children.containsKey(vicinity)) {
+      invokeLayoutCallback<BoxConstraints>((BoxConstraints _) {
+        _childManager.buildChild(vicinity);
+      });
+    } else {
+      _childManager.reuseChild(vicinity);
+    }
+    if (!_children.containsKey(vicinity)) {
+      // There is no child for this vicinity, we may have reached the end of the
+      // children in one or both of the x/y indices.
+      return null;
+    }
+
+    assert(_children.containsKey(vicinity));
+    final RenderBox child = _children[vicinity]!;
+    parentDataOf(child).vicinity = vicinity;
+    return child;
+  }
+
+  /// Called after [layoutChildSequence] to compute the
+  /// [TwoDimensionalViewportParentData.paintOffset] and
+  /// [TwoDimensionalViewportParentData.paintExtent] of the child.
+  void updateChildPaintData(RenderBox child) {
```

Comment:
Wondering why setting these is not just the responsibility of layoutChildSequence